### PR TITLE
mlflow model conversion from HF for image to image task

### PR DESCRIPTION
**Pull request by vadthyavath**
test pipelines: https://ml.azure.com/experiments/id/c41e6aae-f89c-4ccc-9d95-6c6b11572843/runs/tender_dress_w85qhxg5xn?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/rv-images-rc/providers/Microsoft.MachineLearningServices/workspaces/rv-images-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47#

### CODE REVIEW

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 1]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+
+    def __init__(self, **kwargs):
+        """Initialize MLflow convertor for text to image models."""
+        super().__init__(**kwargs)
+
+    def get_model_signature(self):
+        """Return model signature for text to image models."""
+        input_schema = Schema(inputs=[
+            ColSpec(name=TextToImageMLFlowSchemaLiterals.INPUT_COLUMN_PROMPT,
+                    type=TextToImageMLFlowSchemaLiterals.INPUT_COLUMN_PROMPT_DATA_TYPE,),
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:533]:
```suggestion
                    type=TextToImageMLFlowSchemaLiterals.INPUT_COLUMN_PROMPT_DATA_TYPE),
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 2]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+            ColSpec(name=TextToImageMLFlowSchemaLiterals.INPUT_COLUMN_IMAGE,
+                    type=TextToImageMLFlowSchemaLiterals.INPUT_COLUMN_IMAGE_TYPE,)
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:535]:
```suggestion
                    type=TextToImageMLFlowSchemaLiterals.INPUT_COLUMN_IMAGE_TYPE)
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 3]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+        ])
+        output_schema = Schema(inputs=[
+            ColSpec(name=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_IMAGE,
+                    type=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_IMAGE_TYPE),
+            ColSpec(name=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_NSFW_FLAG,
+                    type=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_NSFW_FLAG_TYPE,),
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:541]:
```suggestion
                    type=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_NSFW_FLAG_TYPE),
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 4]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+        ])
+        return ModelSignature(inputs=input_schema, outputs=output_schema)
+
+
+class StableDiffusionImageToImageMlflowConvertor(ImageTextToImageMLflowConvertor):
+    """HF MlfLow convertor class for stable diffusion inpainting models."""
+
+    def __init__(self, **kwargs):
+        """Initialize MLflow convertor for SD inpainting models."""
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:550]:
```suggestion
        """Initialize MLflow convertor for SD image to image models."""
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 5]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+        :return: artifacts dict
+        :rtype: Dict
+        """
+        artifacts_dict = {
+            TextToImageMLflowLiterals.MODEL_DIR: self._model_dir
+        }
+        return artifacts_dict
+
+    def save_as_mlflow(self):
+        """Prepare SD model for save to MLflow."""
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:565]: Don't need two doc strings

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/stable_diffusion_image_to_image_mlflow_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        nsfw_content = None
+
+        df = pd.DataFrame(
+            {
+                MLflowSchemaLiterals.OUTPUT_COLUMN_IMAGE: generated_images,
+                MLflowSchemaLiterals.OUTPUT_COLUMN_NSFW_FLAG: nsfw_content,
+            }
+        )
+
+        return df
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/stable_diffusion_image_to_image_mlflow_wrapper.py:107]: nit: new line at the end of the file.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 7]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+    """MlfLow convertor base class for text to image inpainting models."""
+
+    MODEL_DIR = os.path.join(os.path.dirname(__file__), "text_to_image")
+
+    def __init__(self, **kwargs):
+        """Initialize MLflow convertor for text to image models."""
+        super().__init__(**kwargs)
+
+    def get_model_signature(self):
+        """Return model signature for text to image models."""
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:530]:
```suggestion
        """Return model signature for image-text to image models."""
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 8]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
         )
 
 
+class ImageTextToImageMLflowConvertor(PyFuncMLFLowConvertor):
+    """MlfLow convertor base class for text to image inpainting models."""
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:521]:
```suggestion
    """MlfLow convertor base class for image-text to image models."""
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 9]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+
+    MODEL_DIR = os.path.join(os.path.dirname(__file__), "text_to_image")
+
+    def __init__(self, **kwargs):
+        """Initialize MLflow convertor for text to image models."""
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:526]:
```suggestion
        """Initialize MLflow convertor for image-text to image models."""
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py [hunk 10]
```diff
@@ -517,6 +517,72 @@ def save_as_mlflow(self):
...
+            ColSpec(name=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_IMAGE,
+                    type=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_IMAGE_TYPE),
+            ColSpec(name=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_NSFW_FLAG,
+                    type=TextToImageMLFlowSchemaLiterals.OUTPUT_COLUMN_NSFW_FLAG_TYPE,),
+        ])
+        return ModelSignature(inputs=input_schema, outputs=output_schema)
+
+
+class StableDiffusionImageToImageMlflowConvertor(ImageTextToImageMLflowConvertor):
+    """HF MlfLow convertor class for stable diffusion inpainting models."""
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/convertors.py:547]:
```suggestion
    """HF MlfLow convertor class for stable diffusion image-text to image models."""
```

File: assets/training/model_management/components/convert_model_to_mlflow/spec.yaml [hunk 11]
```diff
@@ -1,7 +1,7 @@
 $schema: https://azuremlschemas.azureedge.net/latest/commandComponent.schema.json
 
 name: convert_model_to_mlflow
-version: 0.0.15
+version: 0.0.17
```
Comment (rjaincc) [inline assets/training/model_management/components/convert_model_to_mlflow/spec.yaml:4]: why are we going from 15 to 17?

File: assets/models/system/stabilityai-stable-diffusion-xl-refiner-1-0/description.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
+___stabilityai/stable-diffusion-xl-refiner-1.0___ employs an ensemble of expert modules in a pipeline for latent diffusion. The process involves using a base model to generate noisy latents, which are then refined using a specialized denoising model. The base model can function independently. Alternatively, a two-stage pipeline involves generating latents with the base model and then refining them using a high-resolution model and the SDEdit technique. The second approach is slightly slower due to more function evaluations.
+
+> The above summary was generated using ChatGPT. Review the <a href="https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0" target="_blank">original-model-card</a> to understand the data used to train the model, evaluation metrics, license, intended uses, limitations and bias before using the model.
```
Comment (rjaincc) [inline assets/models/system/stabilityai-stable-diffusion-xl-refiner-1-0/description.md:3]: The link for HF model card is pointing to the base model, should be pointing to the refiner model

File: assets/models/system/stabilityai-stable-diffusion-xl-refiner-1-0/description.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+### Inference samples
+
+Inference type|Python sample (Notebook)|CLI with YAML
+|--|--|--|
+Real time|<a href="https://aka.ms/azureml-infer-sdk-image-text-to-image-generation" target="_blank">image-text-to-image-online-endpoint.ipynb</a>|<a href="https://aka.ms/azureml-infer-cli-image-text-to-image-generation" target="_blank">image-text-to-image-online-endpoint.sh</a>
```
Comment (rjaincc) [inline assets/models/system/stabilityai-stable-diffusion-xl-refiner-1-0/description.md:9]: Are aka.ms links generated?

**Review by rjaincc — DISMISSED** [review]
Approving with minor comments. Please address these before merging.

File: assets/training/model_management/components/convert_model_to_mlflow/spec.yaml (hunk 11 above)
Comment (vadthyavath) [reply assets/training/model_management/components/convert_model_to_mlflow/spec.yaml:4]: changed to 15!

**Review by MadhuM02 — APPROVED** [review]

**Review by MadhuM02 — DISMISSED** [review]

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/stable_diffusion_image_to_image_mlflow_wrapper.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    get_pil_image,
+    process_image_pandas_series,
+    get_current_device,
+    image_to_base64,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class StableDiffusionImagetoImageMLflowWrapper(mlflow.pyfunc.PythonModel):
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/stable_diffusion_image_to_image_mlflow_wrapper.py:32]: Let's use ImageTexttoImage / image-text-to-image for consistency

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/config.py [hunk 15]
```diff
@@ -18,6 +18,7 @@ class Tasks(_CustomEnum):
 
     TEXT_TO_IMAGE = "text-to-image"
     TEXT_TO_IMAGE_INPAINTING = "text-to-image-inpainting"
+    IMAGE_TO_IMAGE = "image-text-to-image"
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/config.py:21]: I think, would be better to call it "IMAGE_TEXT_TO_IMAGE"

File: assets/models/system/stabilityai-stable-diffusion-xl-refiner-1-0/spec.yaml [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+  inference-min-sku-spec: 4|1|28|176
+  inference-recommended-sku: Standard_NC6s_v3, Standard_NC12s_v3, Standard_NC24s_v3,
+    Standard_NC24rs_v3, Standard_NC16as_T4_v3, Standard_NC24ads_A100_v4, Standard_NC48ads_A100_v4,
+    Standard_NC4as_T4_v3, Standard_NC64as_T4_v3, Standard_NC8as_T4_v3, Standard_NC96ads_A100_v4,
+    Standard_ND40rs_v2, Standard_ND96amsr_A100_v4, Standard_ND96asr_v4
+  model_id: stabilityai/stable-diffusion-xl-refiner-1.0
+tags:
+  Preview: ''
+  license: creativeml-openrail++-m
+  task: image-text-to-image
```
Comment (rjaincc) [inline assets/models/system/stabilityai-stable-diffusion-xl-refiner-1-0/spec.yaml:16]: We want to keep it "image-to-image" --> same as HuggingFace to align

**Review by rjaincc — CHANGES_REQUESTED** [review]
Please address the comments.

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/config.py [hunk 17]
```diff
@@ -18,6 +18,7 @@ class Tasks(_CustomEnum):
 
     TEXT_TO_IMAGE = "text-to-image"
     TEXT_TO_IMAGE_INPAINTING = "text-to-image-inpainting"
+    IMAGE_TEXT_TO_IMAGE = "image-to-image"
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/text_to_image/config.py:21]:
```suggestion
    IMAGE_TO_IMAGE = "image-to-image"
```

File: assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/config.py [hunk 18]
```diff
@@ -40,6 +40,7 @@ class SupportedTasks(_CustomEnum):
     # Text to Image
     TEXT_TO_IMAGE = "text-to-image"
     TEXT_TO_IMAGE_INPAINTING = "text-to-image-inpainting"
+    IMAGE_TEXT_TO_IMAGE = "image-to-image"
```
Comment (rjaincc) [inline assets/training/model_management/src/azureml/model/mgmt/processors/pyfunc/config.py:43]:
```suggestion
    IMAGE_TO_IMAGE = "image-to-image"
```

**Review by MadhuM02 — APPROVED** [review]

**Review by rjaincc — APPROVED** [review]

**Pull request merged**